### PR TITLE
Optimize db config

### LINKED ISSUE #5639: server error sometimes
### Self Checks

- [X] This is only for bug report, if you would like to ask a question, please head to [Discussions](https://github.com/langgenius/dify/discussions/categories/general).
- [X] I have searched for existing issues [search for existing issues](https://github.com/langgenius/dify/issues), including closed ones.
- [X] I confirm that I am using English to submit this report (我已阅读并同意 [Language Policy](https://github.com/langgenius/dify/issues/1542)).
- [X] 请务必使用英文提交 Issue，否则会被关闭。谢谢！:）
- [X] Please do not modify this template :) and fill in all the required fields.

### Dify version

0.6.11

### Cloud or Self Hosted

Self Hosted (Docker)

### Steps to reproduce

I'm using dify with postgresql-ha(bitnami chart) and milvus.
1. Create a text generation workflow.
2. input some message and click generate.

### ✔️ Expected Behavior

No error and with correct response.

### ❌ Actual Behavior

It sometimes failed:
image
Logs:
```
2024-06-26 09:53:13.473 ERROR [Dummy-18] [base_app_generate_response_converter.py:128] - (psycopg2.OperationalError) server closed the connection unexpectedly
This probably means the server terminated abnormally
before or while processing the request.

[SQL: SELECT messages.id AS messages_id, messages.app_id AS messages_app_id, messages.model_provider AS messages_model_provider, messages.model_id AS messages_model_id, messages.override_model_configs AS messages_override_model_configs, messages.conversation_id AS messages_conversation_id, messages.inputs AS messages_inputs, messages.query AS messages_query, messages.message AS messages_message, messages.message_tokens AS messages_message_tokens, messages.message_unit_price AS messages_message_unit_price, messages.message_price_unit AS messages_message_price_unit, messages.answer AS messages_answer, messages.answer_tokens AS messages_answer_tokens, messages.answer_unit_price AS messages_answer_unit_price, messages.answer_price_unit AS messages_answer_price_unit, messages.provider_response_latency AS messages_provider_response_latency, messages.total_price AS messages_total_price, messages.currency AS messages_currency, messages.status AS messages_status, messages.error AS messages_error, messages.message_metadata AS messages_message_metadata, messages.invoke_from AS messages_invoke_from, messages.from_source AS messages_from_source, messages.from_end_user_id AS messages_from_end_user_id, messages.from_account_id AS messages_from_account_id, messages.created_at AS messages_created_at, messages.updated_at AS messages_updated_at, messages.agent_based AS messages_agent_based, messages.workflow_run_id AS messages_workflow_run_id
FROM messages
WHERE messages.id = %(id_1)s::UUID
LIMIT %(param_1)s]
[parameters: {'id_1': 'a3cfcf17-3a26-4766-b1db-197af9ca3515', 'param_1': 1}]
(Background on this error at: https://sqlalche.me/e/20/e3q8)
2024-06-26 10:00:29.163 ERROR [Dummy-27] [app.py:838] - Exception on /console/api/apps/15b513f9-139c-4a17-9792-77ea9833cd18/completion-messages [POST]
Traceback (most recent call last):
File "/usr/local/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1967, in _exec_single_context
self.dialect.do_execute(
File "/usr/local/lib/python3.10/site-packages/sqlalchemy/engine/default.py", line 924, in do_execute
cursor.execute(statement, parameters)
psycopg2.OperationalError: server closed the connection unexpectedly
This probably means the server terminated abnormally
before or while processing the request.


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
File "/usr/local/lib/python3.10/site-packages/flask/app.py", line 880, in full_dispatch_request
rv = self.dispatch_request()
File "/usr/local/lib/python3.10/site-packages/flask/app.py", line 865, in dispatch_request
return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args) # type: ignore[no-any-return]
File "/usr/local/lib/python3.10/site-packages/flask_restful/__init__.py", line 489, in wrapper
resp = resource(*args, **kwargs)
File "/usr/local/lib/python3.10/site-packages/flask/views.py", line 110, in view
return current_app.ensure_sync(self.dispatch_request)(**kwargs) # type: ignore[no-any-return]
File "/usr/local/lib/python3.10/site-packages/flask_restful/__init__.py", line 604, in dispatch_request
resp = meth(*args, **kwargs)
File "/app/api/controllers/console/setup.py", line 83, in decorated
elif not get_setup_status():
File "/app/api/controllers/console/setup.py", line 93, in get_setup_status
return DifySetup.query.first()
File "/usr/local/lib/python3.10/site-packages/sqlalchemy/orm/query.py", line 2728, in first
return self.limit(1)._iter().first() # type: ignore
File "/usr/local/lib/python3.10/site-packages/sqlalchemy/orm/query.py", line 2827, in _iter
result: Union[ScalarResult[_T], Result[_T]] = self.session.execute(
File "/usr/local/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 2351, in execute
return self._execute_internal(
File "/usr/local/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 2236, in _execute_internal
result: Result[Any] = compile_state_cls.orm_execute_statement(
File "/usr/local/lib/python3.10/site-packages/sqlalchemy/orm/context.py", line 293, in orm_execute_statement
result = conn.execute(
File "/usr/local/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1418, in execute
return meth(
File "/usr/local/lib/python3.10/site-packages/sqlalchemy/sql/elements.py", line 515, in _execute_on_connection
return connection._execute_clauseelement(
File "/usr/local/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1640, in _execute_clauseelement
ret = self._execute_context(
File "/usr/local/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1846, in _execute_context
return self._exec_single_context(
File "/usr/local/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1986, in _exec_single_context
self._handle_dbapi_exception(
File "/usr/local/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 2353, in _handle_dbapi_exception
raise sqlalchemy_exception.with_traceback(exc_info[2]) from e
File "/usr/local/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1967, in _exec_single_context
self.dialect.do_execute(
File "/usr/local/lib/python3.10/site-packages/sqlalchemy/engine/default.py", line 924, in do_execute
cursor.execute(statement, parameters)
sqlalchemy.exc.OperationalError: (psycopg2.OperationalError) server closed the connection unexpectedly
This probably means the server terminated abnormally
before or while processing the request.

[SQL: SELECT dify_setups.version AS dify_setups_version, dify_setups.setup_at AS dify_setups_setup_at
FROM dify_setups
LIMIT %(param_1)s]
[parameters: {'param_1': 1}]
(Background on this error at: https://sqlalche.me/e/20/e3q8)
2024-06-26 10:15:33.575 ERROR [Dummy-51] [app.py:838] - Exception on /api/completion-messages [POST]
Traceback (most recent call last):
File "/usr/local/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1967, in _exec_single_context
self.dialect.do_execute(
File "/usr/local/lib/python3.10/site-packages/sqlalchemy/engine/default.py", line 924, in do_execute
cursor.execute(statement, parameters)
psycopg2.OperationalError: server closed the connection unexpectedly
This probably means the server terminated abnormally
before or while processing the request.


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
File "/usr/local/lib/python3.10/site-packages/flask/app.py", line 880, in full_dispatch_request
rv = self.dispatch_request()
File "/usr/local/lib/python3.10/site-packages/flask/app.py", line 865, in dispatch_request
return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args) # type: ignore[no-any-return]
File "/usr/local/lib/python3.10/site-packages/flask_restful/__init__.py", line 489, in wrapper
resp = resource(*args, **kwargs)
File "/usr/local/lib/python3.10/site-packages/flask/views.py", line 110, in view
return current_app.ensure_sync(self.dispatch_request)(**kwargs) # type: ignore[no-any-return]
File "/usr/local/lib/python3.10/site-packages/flask_restful/__init__.py", line 604, in dispatch_request
resp = meth(*args, **kwargs)
File "/app/api/controllers/web/wraps.py", line 18, in decorated
app_model, end_user = decode_jwt_token()
File "/app/api/controllers/web/wraps.py", line 45, in decode_jwt_token
app_model = db.session.query(App).filter(App.id == decoded['app_id']).first()
File "/usr/local/lib/python3.10/site-packages/sqlalchemy/orm/query.py", line 2728, in first
return self.limit(1)._iter().first() # type: ignore
File "/usr/local/lib/python3.10/site-packages/sqlalchemy/orm/query.py", line 2827, in _iter
result: Union[ScalarResult[_T], Result[_T]] = self.session.execute(
File "/usr/local/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 2351, in execute
return self._execute_internal(
File "/usr/local/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 2236, in _execute_internal
result: Result[Any] = compile_state_cls.orm_execute_statement(
File "/usr/local/lib/python3.10/site-packages/sqlalchemy/orm/context.py", line 293, in orm_execute_statement
result = conn.execute(
File "/usr/local/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1418, in execute
return meth(
File "/usr/local/lib/python3.10/site-packages/sqlalchemy/sql/elements.py", line 515, in _execute_on_connection
return connection._execute_clauseelement(
File "/usr/local/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1640, in _execute_clauseelement
ret = self._execute_context(
File "/usr/local/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1846, in _execute_context
return self._exec_single_context(
File "/usr/local/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1986, in _exec_single_context
self._handle_dbapi_exception(
File "/usr/local/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 2353, in _handle_dbapi_exception
raise sqlalchemy_exception.with_traceback(exc_info[2]) from e
File "/usr/local/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1967, in _exec_single_context
self.dialect.do_execute(
File "/usr/local/lib/python3.10/site-packages/sqlalchemy/engine/default.py", line 924, in do_execute
cursor.execute(statement, parameters)
sqlalchemy.exc.OperationalError: (psycopg2.OperationalError) server closed the connection unexpectedly
This probably means the server terminated abnormally
before or while processing the request.

[SQL: SELECT apps.id AS apps_id, apps.tenant_id AS apps_tenant_id, apps.name AS apps_name, apps.description AS apps_description, apps.mode AS apps_mode, apps.icon AS apps_icon, apps.icon_background AS apps_icon_background, apps.app_model_config_id AS apps_app_model_config_id, apps.workflow_id AS apps_workflow_id, apps.status AS apps_status, apps.enable_site AS apps_enable_site, apps.enable_api AS apps_enable_api, apps.api_rpm AS apps_api_rpm, apps.api_rph AS apps_api_rph, apps.is_demo AS apps_is_demo, apps.is_public AS apps_is_public, apps.is_universal AS apps_is_universal, apps.created_at AS apps_created_at, apps.updated_at AS apps_updated_at
FROM apps
WHERE apps.id = %(id_1)s::UUID
LIMIT %(param_1)s]
[parameters: {'id_1': '15b513f9-139c-4a17-9792-77ea9833cd18', 'param_1': 1}]
(Background on this error at: https://sqlalche.me/e/20/e3q8)
```

### PR DESCRIPTION
**Pull request by NiuBlibing**
# Checklist:

> [!IMPORTANT]  
> Please review the checklist below before submitting your pull request.

- [x] Please open an issue before creating a PR or link to an existing issue
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I ran `dev/reformat`(backend) and `cd web && npx lint-staged`(frontend) to appease the lint gods

# Description

1. support [sqlalchemy debug mode](https://docs.sqlalchemy.org/en/20/core/engines.html#sqlalchemy.create_engine.params.echo)
2. replace `DB_CHARSET` config with `DB_EXTRAS` which is more flexible.

Fixes https://github.com/langgenius/dify/issues/5639

## Type of Change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update, included: [Dify Document](https://github.com/langgenius/dify-docs)
- [ ] Improvement, including but not limited to code refactoring, performance optimization, and UI/UX improvement
- [ ] Dependency upgrade

# Testing Instructions

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

### CODE REVIEW

File: api/configs/middleware/__init__.py [hunk 1]
```diff
@@ -89,7 +89,7 @@ class DatabaseConfig:
     @computed_field
     @property
     def SQLALCHEMY_DATABASE_URI(self) -> str:
-        db_extras = f"?client_encoding={self.DB_CHARSET}" if self.DB_CHARSET else ""
+        db_extras = f"?{self.DB_EXTRAS}" if self.DB_EXTRAS else ""
```
Comment (takatost) [inline api/configs/middleware/__init__.py]: This change will result in a breaking change, and we need to keep the previous DB_CHARSET. Ensure forward compatibility

**Review by takatost — CHANGES_REQUESTED** [review]

File: api/configs/middleware/__init__.py (hunk 1 above)
Comment (NiuBlibing) [reply api/configs/middleware/__init__.py]: Fixed

File: api/configs/middleware/__init__.py [hunk 2]
```diff
@@ -81,6 +81,11 @@ class DatabaseConfig:
         default='',
     )
 
+    DB_EXTRAS: str = Field(
+        iption='db extras options. Example: keepalives_idle=60&keepalives=1',
```
Comment (takatost) [inline api/configs/middleware/__init__.py]: Should the name of the parameter here be `description`?

**Review by takatost — CHANGES_REQUESTED** [review]

File: api/configs/middleware/__init__.py (hunk 2 above)
Comment (NiuBlibing) [reply api/configs/middleware/__init__.py]: fixed

**Review by crazywoola — APPROVED** [review]

**Review by takatost — APPROVED** [review]
thanks!

**Pull request merged**